### ISSUE TITLE
Prune search space using 'deadend' edge labels for faster flexible routing

### ISSUE THREAD

**Issue by easbar** [graphhopper/graphhopper#2076]
Valhalla is using a special optimization to prune the shortest path search space. Edges that only give access to 'dead end' areas but are normally not required to connect points from different areas are not expanded during the search. In the most simple example the search does not enter a dead end road unless the target lies on this road, but more generally there can be entire 'subtrees' that are only connected by a single edge to the main graph and can be skipped if neither start nor end points lie within them. In case the start/target point lies within such a 'dead end' area the routing is still correct as long as we use a bidirectional search. For example, if the target point lies in such a dead end area the forward search will not enter it, but the backward search will still be able to leave this area. Valhalla calls such edges 'no thru' edges and they are labeled during graph import (finding these edges requires extra graph processing). I did a quick test for an example route and compared the number of Dijkstra loop iterations with and without this optimization:

|vehicle|iterations without optimization|iterations with optimization|
|-|-|-|
|car|45.304|45.031|
|bicycle|4.558.470|3.915.926|
|pedestrian|1.412.917|1.205.452|

It can be seen that for bicycle and pedestrian routing the number of Dijkstra iterations is reduced by about 15%. For car routing the effect is much smaller because Valhalla's road hierarchy drastically reduces the search space size already (this technique is not used for bicycle and pedestrian routing). The query time was roughly proportional to the number of Dijkstra iterations I measured. Note that Valhalla's car/bicycle routing is much slower than for cars.

A map showing the dead end edges found by Valhalla looks like this:

![image](https://user-images.githubusercontent.com/17603532/86532841-2fef3380-becd-11ea-9849-2fa6bc4fed3e.png)

There are mostly simple dead ends (which probably do not reduce the search space size very much), but sometimes there are bigger subtrees as well.

I would expect that this optimization might yield a 10-15% performance improvement for GraphHopper's flex routing (non-CH/LM) and maybe its worth a try. For LM it might be even more useful because in this case the absolute search space size is already quite small and the relative improvement might be bigger (but this is just a guess).

**Comment by michaz** [comment]
As always (boring, I know), I think code should not be written, and concepts should not be introduced, for 10-15% performance improvements. If a 10-15% performance improvement can be achieved by removing or cleaning stuff: Great investment.

**Comment by michaz** [comment]
Still interesting though!

**Comment by easbar** [comment]
Yes, initially I was hoping it would give a better speed up and did the experiments I described above. If the improvement is only like 10-15% I also think its probably not worth it, but this was only a very rough estimation and maybe I did not even do it the right way, so my conclusion was: 'maybe try it again at some point, but not for now'.
